### DISCUSSION TITLE
Questions about Key Mapping, Model Name Display, and Autocompletion for CodeCompanion Command

### DISCUSSION THREAD

**Discussion by bassamsdata** [olimorris/codecompanion.nvim#285] [Q&A] [answered]
Hello,

While maximizing my usage of the plugin, I've run into a few questions.

1. When I map the `CodeCompanion` command to a keymap, it doesn't seem to work. However, it works fine when manually typed in the command line. Here's the mapping I'm using:
   ```lua
   vim.keymap.set({ "n", "v" }, "<leader>ai", "<cmd>CodeCompanion<cr>", { desc = "[A]i [I]nline" })
   ```

2. I'd like to display the model name (not the provider) that outputs the data beneath the `CodeCompanion` text. However, I'm unsure where to retrieve this information. The `CodeCompanionChatAdapter` event only includes the default model, and I couldn't find anything related to dynamic model names.

   The reason I'm asking is that I want to avoid displaying the settings at the top of the chat buffer when it grows too long. This would reduce key strokes, and since I'm experimenting with multiple models at the moment, it would help streamline things. (Apologies for nitpicking on this.)

3. Is it possible to add autocompletions for the `CodeCompanionChat` command? It has quite a few subcommands now, and I'd find this very useful. I wanted to submit a PR for it, but I've never worked with completion before.

Thank you!

**Comment by S1M0N38** [comment]
hi,

1. Did you try to set up your keymap in minimal.lua config in the issue template (you can get that when opening an issue)? maybe you issue is caused by keymaps overlapping.
2. idk. sorry
3. Check out https://github.com/nvim-neorocks/nvim-best-practices?tab=readme-ov-file#speaking_head-user-commands . Here they explain how completion for subcommands can be implemented. For a simplify version you can check my [nvim plugin template](https://github.com/S1M0N38/base.nvim/blob/main/plugin/base.lua)

**Comment by olimorris** [comment, marked as answer]
Regarding 1), that's because the command expects some input before it can run. I recommend in the workflow tips to apply an abbreviation such as `cc` as it's 3 keystrokes before you're typing a prompt. Alternatively, you could use the prompt library shortcut for the custom prompt.

For item 2), this is a little confusing as `default_model` is poorly named. It can represent the default and the current model. The payload sent from the `CodeCompanionChatAdapter` event will enable you to access this. Another user had a similar request recently which is how it came to be included in the plugin.

And great shout on 3). Thanks to @S1M0N38 as I hadn't realised you could do this with sub commands.

**Comment by bassamsdata** [comment]
Thank you all for the responses! the plugin template and the guide for best practices look amazing and easy to follow. I'll give it a try and see if I can get the autocompletion for the subcommands working.

@olimorris, are you talking about `schema.model.default`? If so, that always points to the default model in the plugin config and doesn’t update to reflect the current model in use (though the adapter name does change).

**Reply by olimorris** [reply]
Ooooh you're right. I'll add a new command that fires the new model!

**Reply by olimorris** [reply]
@bassamsdata this should now be added in `5.3.0`

**Reply by olimorris** [reply]
Regarding subcommands, if we were to add them, we'd need to add the slash commands from the prompt_library for full coverage. I'm not convinced their worthwhile adding in. We'll also need to make sure we don't break the `legendary.nvim` config.

**Comment by bassamsdata** [comment]
@olimorris Thanks for the quick implementation—it works perfectly!

> if we were to add them, we'd need to add the slash commands. We'll also need to make sure we don't break the legendary.nvim config.

Sure, makes sense.

> I'm not convinced they're worthwhile adding in.

I'll give it a go and see what I can come up with. If it works, awesome! If not, at least I’ll have learned something new. Plus, I'm working on a module for yanking, so I’m thinking of adding completions for that as well.

**Reply by olimorris** [reply]
Sounds good, feel free to open up a PR to discuss it if it's going to change the UI or alter the API

**Comment by olimorris** [comment]
I've added in sub commands as of `8.1.0`

**Reply by bassamsdata** [reply]
that's great, thank you

**Reply by bassamsdata** [reply]
Just a small nitpick: the completion for both adapters and custom prompts doesn’t seem to recognize user-defined adapters or prompts. After checking the code, it looks like it *should* be picking them up, since the setup function updates the config—but I’m not sure why it’s not working as expected.

**Reply by olimorris** [reply]
I tested it with mine and it seems to work. I did push that fix late into the day yesterday so are you on the latest version of the plugin?

**Reply by bassamsdata** [reply]
Looks like I commented just after the fix went live. It’s working perfectly now. thanks
